pkg/controller/linstornodeset/linstornodeset_controller.go

```diff
@@ -329,180 +334,283 @@ func (r *ReconcileLinstorNodeSet) reconcileResource(ctx context.Context, pns *pi
+func (r *ReconcileLinstorNodeSet) reconcileSingleNodeRegistration(ctx context.Context, nodeSet *piraeusv1alpha1.LinstorNodeSet, pod *corev1.Pod) error {
+	lNode, err := r.linstorClient.GetNodeOrCreate(ctx, lapi.Node{
 		Name: pod.Spec.NodeName,
 		Type: lc.Satellite,
 		NetInterfaces: []lapi.NetInterface{
 			{
 				Name:                    "default",
 				Address:                 pod.Status.HostIP,
-				SatellitePort:           ssl.Port(),
-				SatelliteEncryptionType: ssl.Type(),
+				SatellitePort:           nodeSet.Spec.SslConfig.Port(),
+				SatelliteEncryptionType: nodeSet.Spec.SslConfig.Type(),
 			},
 		},
 	})
 	if err != nil {
-		log.Debug("NS reconcileSatNodeWithController error")
 		return err
 	}
 
-	log.WithFields(logrus.Fields{
-		"nodeName":         node.Name,
-		"nodeType":         node.Type,
-		"connectionStatus": node.ConnectionStatus,
-	}).Debug("NS reconcileSatNodeWithController: Found / Added a Satellite Node")
-
-	sat.ConnectionStatus = node.ConnectionStatus
-	if sat.ConnectionStatus != lc.Online {
-		return fmt.Errorf("NS reconcileSatNodeWithController: waiting for node %s ConnectionStatus to be %s, current ConnectionStatus: %s",
-			pod.Spec.NodeName, lc.Online, sat.ConnectionStatus)
+	if lNode.ConnectionStatus != lc.Online {
+		return fmt.Errorf("node '%s' registered, but not online (%s)", lNode.Name, lNode.ConnectionStatus)
 	}
 
-	sat.RegisteredOnController = true
 	return nil
 }
 
-func (r *ReconcileLinstorNodeSet) reconcileStoragePoolsOnNode(sat *piraeusv1alpha1.SatelliteStatus, pools []piraeusv1alpha1.StoragePool, pod corev1.Pod) error {
+func (r *ReconcileLinstorNodeSet) reconcileAutomaticDeviceSetup(ctx context.Context, nodeSet *piraeusv1alpha1.LinstorNodeSet, pod *corev1.Pod) error {
+	logger := logrus.WithFields(logrus.Fields{
+		"Name":      nodeSet.Name,
+		"Namespace": nodeSet.Namespace,
+		"Pod":       pod.Name,
+		"Op":        "reconcileAutomaticDeviceSetup",
+	})
+	// No setup required
+	if nodeSet.Spec.AutomaticStorageSetupType == defaultAutomaticStorageSetupType {
+		return nil
+	}
+
+	logger.Warn("Automatic device set up enabled. This is a potentially dangerous operation!")
+
+	logger.Debug("fetch available devices for node")
+
+	storageList, err := r.linstorClient.Nodes.GetPhysicalStorage(ctx, &lapi.ListOpts{Node: []string{pod.Spec.NodeName}})
+	if err != nil {
+		return err
+	}
+
+	paths := make([]string, 0)
+
+	for _, entry := range storageList {
+		ourNode, ok := entry.Nodes[pod.Spec.NodeName]
+		if !ok {
+			continue
+		}
+
+		for _, dev := range ourNode {
+			paths = append(paths, dev.Device)
+		}
+	}
+
+	// LINSTOR does not like an empty device path list
+	if len(paths) == 0 {
+		logger.Debug("no devices left to set up")
+		return nil
+	}
+
+	logger.WithField("devices", paths).Debug("use devices for storage pool")
+
+	return r.linstorClient.Nodes.CreateDevicePool(ctx, pod.Spec.NodeName, lapi.PhysicalStorageCreate{
+		DevicePaths: paths,
```

Comment:
We may actually want one storage pool per device rather than one for all devices. It is a bit awkward.

One pool per device advantages:
* Better availability - if one device goes down fewer volumes affected
* More flexibility - we don't know if we are combining completely different types of devices here

One for all devices advantages:
* More efficient use of space because we don't have to have space on a single device for a volume

We should get other opinions on this.